sys/aarch64/pl011.c

```diff
@@ -57,103 +76,130 @@ static uint32_t pl011_getc(pl011_state_t *state) {
+      while (pl011_wready(pl011) && ringbuf_getb(&pl011->tx_buf, &byte)) {
+        pl011_putc(pl011, byte);
       }
+      if (ringbuf_empty(&pl011->tx_buf)) {
+        /* If we're out of characters and there are characters
+         * in the tty's output queue, signal the tty thread to refill. */
+        if (pl011->tty_thread_flags & TTY_THREAD_OUTQ_NONEMPTY) {
+          pl011->tty_thread_flags |= TTY_THREAD_TXRDY;
+          cv_signal(&pl011->tty_thread_cv);
+        }
+        /* Disable TXRDY interrupts - the tty thread will re-enable them
+         * after filling tx_buf. */
+        clr4(pl011->regs, PL011COM_CR, PL011_CR_TXE);
+      }
+      res = IF_FILTERED;
     }
   }
 
-  return 0;
+  return res;
 }
 
-static int pl011_close(vnode_t *v, file_t *fp) {
-  /* TODO release resources */
-  return 0;
-}
+/*
+ * If tx_buf is empty, we can try to write characters directly from tty->t_outq.
+ * This routine attempts to do just that.
+ * Must be called with both tty->t_lock and pl011->lock held.
+ */
+static void pl011_try_bypass_txbuf(pl011_state_t *pl011, tty_t *tty) {
+  uint8_t byte;
 
-static int pl011_ioctl(vnode_t *v, u_long cmd, void *data) {
-  if (cmd) {
-    unsigned len = IOCPARM_LEN(cmd);
-    memset(data, 0, len);
-    return 0;
-  }
+  if (!ringbuf_empty(&pl011->tx_buf))
+    return;
 
-  return EPASSTHROUGH;
+  while (pl011_wready(pl011) && ringbuf_getb(&tty->t_outq, &byte)) {
+    pl011_putc(pl011, byte);
+  }
 }
 
-static int pl011_getattr(vnode_t *v, vattr_t *va) {
-  memset(va, 0, sizeof(vattr_t));
-  va->va_mode = S_IFCHR;
-  va->va_nlink = 1;
-  va->va_ino = 0;
-  va->va_size = 0;
-  return 0;
+static void pl011_set_tty_outq_nonempty_flag(pl011_state_t *pl011, tty_t *tty) {
+  if (ringbuf_empty(&tty->t_outq))
+    pl011->tty_thread_flags &= ~TTY_THREAD_OUTQ_NONEMPTY;
+  else
+    pl011->tty_thread_flags |= TTY_THREAD_OUTQ_NONEMPTY;
 }
 
-/* clang-format off */
-static vnodeops_t dev_uart_ops = {
-  .v_write = pl011_write,
-  .v_read = pl011_read,
-  .v_close = pl011_close,
-  .v_ioctl = pl011_ioctl,
-  .v_getattr = pl011_getattr
-};
-/* clang-format on */
-
-static intr_filter_t pl011_intr(void *data /* device_t* */) {
-  pl011_state_t *state = ((device_t *)data)->state;
-  intr_filter_t res = IF_STRAY;
+/*
+ * Move characters from tty->t_outq to pl011->tx_buf.
+ * Must by called with tty->t_lock held.
+ */
+static void pl011_fill_txbuf(pl011_state_t *pl011, tty_t *tty) {
+  uint8_t byte;
 
-  WITH_SPIN_LOCK (&state->lock) {
-    if (pl011_rready(state)) {
-      ringbuf_putb(&state->rx_buf, pl011_getc(state));
-      cv_signal(&state->rx_nonempty);
-      res = IF_FILTERED;
+  while (true) {
+    SCOPED_SPIN_LOCK(&pl011->lock);
+    pl011_try_bypass_txbuf(pl011, tty);
+    if (ringbuf_full(&pl011->tx_buf) || !ringbuf_getb(&tty->t_outq, &byte)) {
+      pl011_set_tty_outq_nonempty_flag(pl011, tty);
+      break;
     }
+    if (!ringbuf_empty(&pl011->tx_buf))
+      set4(pl011->regs, PL011COM_CR, PL011_CR_TXE);
```

Comment:
Is there a particular reason you moved these 2 lines here? In the `ns16550` version, they are inside the `if` above.

Reply:
It's my fault. Fixed.